group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2427,25 +2471,259 @@ private boolean hasMemberSubscriptionChanged(
+                    // We also trigger a refresh of the regexes in order to resolve it.
                     throwIfRegularExpressionIsInvalid(updatedMember.subscribedTopicRegex());
+                    requireRefresh = true;
+                } else {
+                    // If the new regex is already resolved, we trigger a rebalance
+                    // by bumping the group epoch.
+                    bumpGroupEpoch = group.resolvedRegularExpression(newSubscribedTopicRegex).isPresent();
                 }
-                return true;
             }
         }
-        return false;
+
+        String key = group.groupId() + "-regex";
+
+        // If a refresh is already inflight, we can schedule a new one so
+        // we stop here.
+        if (executor.isScheduled(key)) {
+            return bumpGroupEpoch;
+        }
+
+        // Adjust the count to account for the updated member.
+        Map<String, Integer> subscribedRegularExpressions = new HashMap<>(group.subscribedRegularExpressions());
+        if (oldSubscribedTopicRegex != null) {
+            subscribedRegularExpressions.compute(oldSubscribedTopicRegex, Utils::decValue);
+        }
+        if (newSubscribedTopicRegex != null) {
+            subscribedRegularExpressions.compute(newSubscribedTopicRegex, Utils::incValue);
+        }
+
+        // If the number of subscribed regexes is different from the number of resolved ones,
+        // we must trigger a refresh.
+        requireRefresh |= subscribedRegularExpressions.size() != group.numResolvedRegularExpressions();
+
+        // If the last time that regexes were refreshed is older than the refresh interval,
+        // we must trigger a refresh.
+        requireRefresh |= time.milliseconds() >
+            group.lastResolvedRegularExpressionRefreshTimeMs() + consumerGroupRegularExpressionRefreshInternalMs;
+
+        if (requireRefresh) {
+            Set<String> regexes = subscribedRegularExpressions.keySet();
+            executor.schedule(
+                key,
+                () -> refreshRegularExpressions(groupId, log, time, metadataImage, regexes),
+                (result, exception) -> handleRegularExpressionsResult(groupId, result, exception)
+            );
+        }
+
+        return bumpGroupEpoch;
+    }
+
+    /**
+     * Resolves the provided regular expressions. Note that this static method is executed
+     * as an asynchronous task in the executor. Hence, it should not access any state from
+     * the manager.
+     *
+     * @param groupId   The group id.
+     * @param log       The log instance.
+     * @param time      The time instance.
+     * @param image     The metadata image to use for listing the topics.
+     * @param regexes   The list of regular expressions that must be resolved.
+     * @return The list of resolved regular expressions.
+     */
+    private static Map<String, ResolvedRegularExpression> refreshRegularExpressions(
```

Comment:
yeah, i was considering this too. i was wondering whether it would be too spammy if they are a lot of regexes in the system. what do you think?

Reply:
I'm a bit confused. Isn't this method invoked for each group, and not regex?

I agree it can get spammy. What if we record it in the HdrHistogram? Maybe we can create a separate jira for this?